### ISSUE TITLE
Obtaining the "input" results for the voxel use case

### ISSUE THREAD

**Issue by noamgat** [autonomousvision/occupancy_networks#31]
Hi,

I'm doing followup research on the voxels use case of this paper, and am trying to reproduce the results of the paper before continuing.

I install the environment on ubuntu, and run the following:

python eval.py configs/voxels/onet_pretrained.yml

I obtain the following results (ran once on chairs dataset only, and once on everything):

Chairs Only
                 iou  iou_voxels   kl      loss  rec_error
class name
n/a         0.663234    0.659272  0.0  81.09695   81.09695
mean        0.663234    0.659272  0.0  81.09695   81.09695

Everything
                 iou  iou_voxels   kl      loss  rec_error
class name
n/a         0.695912     0.68121  0.0  57.70389   57.70389
mean        0.695912     0.68121  0.0  57.70389   57.70389

What is the difference between iou and iou_voxels? 
Paper says 
Input IOU 0.631
ONet IOU 0.703

How was the 0.631 obtained, and is it possible to reproduce the ONet IOU with the supplied code?

**Comment by LMescheder** [comment]
Hi @noamgat,
iou is the continuous IOU wrt the GT mesh, while iou_voxels is the iou wrt voxelized mesh when you naively voxelize our prediction at resolution 32^3 (i.e. evaluate the network in 32^3 grid). iou_voxels should only be taken as a rough hint if the model is doing what it is supposed to do, not as an evaluation metric. See [here]( https://github.com/autonomousvision/occupancy_networks/blob/master/im2mesh/onet/training.py#L53) for implementation details

For the evaluation in the paper we actually do not use `eval.py` (which is for quick & dirty evaluation), but instead `eval_meshes.py`, which evaluates the predicted meshes in a standardized way.

To evaluate the input meshes, you have to first generate the meshes and then run `eval_meshes.py` with the `--eval_input` flag.

**Comment by noamgat** [comment]
Thanks for the answer!
So the way to generate the results would be:

python generate.py configs/voxels/onet_pretrained.yaml
python eval_meshes.py configs/voxels/onet_pretrained.yaml

and compare that to

python eval_meshes.py --eval_input configs/voxels/onet_pretrained.yaml

If so, I am getting a lot of "Warning: contains1 != contains2 for some points." during the eval_meshes runs, and then the results look meaningless. The run of generate.py seemed clean.

However, the generated meshes do look reasonable.
[cc1b4eb1a9164e04f06885bd08de3f64.zip](https://github.com/autonomousvision/occupancy_networks/files/3776046/cc1b4eb1a9164e04f06885bd08de3f64.zip)

Any idea on what could be going wrong?

**Comment by LMescheder** [comment]
@noamgat Ok, I will have a look into the issue. Thanks for reporting it!

**Comment by LMescheder** [comment]
@noamgat Just wanted to confirm that I can reproduce the problem. Somehow, this only happens when I run the code on the compressed data that we uploaded, but does not happen when I use our internal, uncompressed dataset (which we used to produce the results in the paper).
Maybe something went wrong when I compressed the dataset. Might take me some time to fix the problem, but I am at it.

**Comment by noamgat** [comment]
Thank you very much for the response and attention!
If there is anything I can do to help, let me know.

**Comment by LMescheder** [comment]
Fixed in 0c44de4e58ebf6119060248e1e35484197947797. I now get IOU of 0.63099  for the input and 0.70344 for output which is identical with what we reported in the paper.
When debugging I also found that the `metadata.yaml` was missing in the dataset we released, so there results were not sorted by class. I will push the updated dataset later.

**Comment by LMescheder** [comment]
#34

**Comment by noamgat** [comment]
I can confirm that the situation is much better. I now have 0.620 for the input and 0.696 for the pretrained model. While not 100% the same as what you are reporting, this is much better than before.

**Comment by LMescheder** [comment]
@noamgat I believe this is due to the way how the results are averaged. In the paper we compute all statistics per class and then average over all classes. Because of the missing `metadata.yaml` file (issue: #34), the test code averaged over all models directly. I just pushed the updated dataset, the error should not occur anymore. If you do not want to download the whole dataset again, you can just copy the `data/metadata.yaml` file into `data/ShapeNet`.
